### ISSUE TITLE
(aws-fis): MapProperties generate empty CFN output

### ISSUE THREAD

**Issue by jwoehrle** [aws/aws-cdk#14309]
I'm trying to generate a `CfnExperimentTemplate` with an `CfnExperimentTemplate.ExperimentTemplateActionProperty` and a `CfnExperimentTemplate.ExperimentTemplateTargetProperty`.  I define it as follows:
```
const action: fis.CfnExperimentTemplate.ExperimentTemplateActionProperty = {
      actionId: 'aws:ec2:stop-instances',
      parameters: { startInstancesAfterDuration: 'PT1M' },
      targets: { Instances: 'myInstance'}
    }
const target: fis.CfnExperimentTemplate.ExperimentTemplateTargetProperty = {
      resourceType: 'aws:ec2:instance',
      resourceTags: { env: 'DEV' },
      selectionMode: 'COUNT(1)'
    }
const template = new fis.CfnExperimentTemplate(this,'fis-template', {
      description: 'My first Template',
      roleArn: role.roleArn,
      stopConditions: [
        { source: 'none' }
      ],
      tags: {},
      actions: {
        'TerminateInstances' : action
      },
      targets: {
        'myInstance': target
      }
    });
```

when doing a `cdk synth` this results in the following

```
Type: AWS::FIS::ExperimentTemplate
    Properties:
      description: My first Template
      roleArn:
        Fn::GetAtt:
          - fisrole33E76559
          - Arn
      stopConditions:
        - source: none
      tags: {}
      targets:
        myInstance:
          resourceTags: {}
          resourceType: aws:ec2:instance
          selectionMode: COUNT(1)
      actions:
        TerminateInstances:
          actionId: aws:ec2:stop-instances
          parameters: {}
          targets: {}
```

note the empty `resourceTags` in the `targets` as well as several other properties being empty (e.g. `actions.TerminateInstance.targets`)

### What did you expect to happen?

The properties I define in CDK should all show up in CFN.

### Environment

  - **CDK CLI Version  :**
  1.100.0
  - **Framework Version:**
  1.100.0
  - **Node.js Version:** 
   14.16.1
  - **OS               :**
  macOs Big Sur
  - **Language (Version):** 
    TypeScript

--- 

This is :bug: Bug Report

**Comment by peterwoodworth** [comment]
I can confirm I can reproduce this behavior

**Comment by peterwoodworth** [comment]
This issue is still valid even with the merging of the new spec. The resource spec defines your missing types like so, which is why they are still empty:

```json
    "AWS::FIS::ExperimentTemplate.TagMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-tagmap.html"
    },
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateActionItemTargetMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateactionitemtargetmap.html"
    },
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateActionItemParameterMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateactionitemparametermap.html"
    },
```

**Comment by skinny85** [comment]
@peterwoodworth the types you quoted do not matter that they are empty - they are never used in the current spec (feel free to confirm by searching [here](https://raw.githubusercontent.com/aws/aws-cdk/acc7bdd6445dde1f49a6379e4a4e9be6bb826e23/packages/%40aws-cdk/cfnspec/spec-source/000_CloudFormationResourceSpecification.json)).

**Comment by skinny85** [comment]
Relevant parts of the spec are:

```json
    "AWS::FIS::ExperimentTemplate": {
      "Attributes": {
        "Id": {
          "PrimitiveType": "String"
        }
      },
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html",
      "Properties": {
        "Actions": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html#cfn-fis-experimenttemplate-actions",
          "ItemType": "ExperimentTemplateAction",
          "Required": false,
          "Type": "Map",
          "UpdateType": "Mutable"
        },
        "Description": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html#cfn-fis-experimenttemplate-description",
          "PrimitiveType": "String",
          "Required": true,
          "UpdateType": "Mutable"
        },
        "RoleArn": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html#cfn-fis-experimenttemplate-rolearn",
          "PrimitiveType": "String",
          "Required": true,
          "UpdateType": "Mutable"
        },
        "StopConditions": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html#cfn-fis-experimenttemplate-stopconditions",
          "ItemType": "ExperimentTemplateStopCondition",
          "Required": true,
          "Type": "List",
          "UpdateType": "Mutable"
        },
        "Tags": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html#cfn-fis-experimenttemplate-tags",
          "PrimitiveItemType": "String",
          "Required": true,
          "Type": "Map",
          "UpdateType": "Immutable"
        },
        "Targets": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html#cfn-fis-experimenttemplate-targets",
          "ItemType": "ExperimentTemplateTarget",
          "Required": true,
          "Type": "Map",
          "UpdateType": "Mutable"
        }
      }
    },
```

**Comment by skinny85** [comment]
```json
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateAction": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html",
      "Properties": {
        "ActionId": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-actionid",
          "PrimitiveType": "String",
          "Required": true,
          "UpdateType": "Mutable"
        },
        "Description": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-description",
          "PrimitiveType": "String",
          "Required": false,
          "UpdateType": "Mutable"
        },
        "Parameters": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-parameters",
          "Required": false,
          "Type": "ExperimentTemplateActionItemParameterMap",
          "UpdateType": "Mutable"
        },
        "StartAfter": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-startafter",
          "PrimitiveItemType": "String",
          "Required": false,
          "Type": "List",
          "UpdateType": "Mutable"
        },
        "Targets": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-targets",
          "Required": false,
          "Type": "ExperimentTemplateActionItemTargetMap",
          "UpdateType": "Mutable"
        }
      }
    },
```

**Comment by skinny85** [comment]
Ah, I see now. The problem are the `Parameters` and `Targets` properties of the `AWS::FIS::ExperimentTemplate.ExperimentTemplateAction` type.

**Comment by skinny85** [comment]
Apologies @peterwoodworth , you were right.

**Comment by peterwoodworth** [comment]
Yes I was just gonna correct ya on that

**Comment by skinny85** [comment]
@peterwoodworth since this is a mistake in the Resource Specification, we need to contact the CloudFormation team about this, giving them the info from https://github.com/aws/aws-cdk/issues/14309#issuecomment-832253367 and https://github.com/aws/aws-cdk/issues/14309#issuecomment-832102006.

**Comment by skinny85** [comment]
I believe the correct specification should be:

```json
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateAction": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html",
      "Properties": {
        "Parameters": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-parameters",
          "Required": false,
          "Type": "Map",
          "ItemType": "ExperimentTemplateActionItemParameter",
          "UpdateType": "Mutable"
        },
        "Targets": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-targets",
          "Required": false,
          "Type": "Map",
          "ItemType": "ExperimentTemplateActionItemTarget",
          "UpdateType": "Mutable"
        }
```

Note that `AWS::FIS::ExperimentTemplate.ExperimentTemplateActionItemTarget` and `AWS::FIS::ExperimentTemplate.ExperimentTemplateActionItemParameter ` are not in the Resource Specification at all.

**Comment by skinny85** [comment]
Small correction - looks like this should be actually:

```json
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateAction": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html",
      "Properties": {
        "Parameters": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-parameters",
          "Required": false,
          "Type": "Map",
          "PrimitiveItemType": "String",
          "UpdateType": "Mutable"
        },
        "Targets": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateaction.html#cfn-fis-experimenttemplate-experimenttemplateaction-targets",
          "Required": false,
          "Type": "Map",
          "PrimitiveItemType": "String",
          "UpdateType": "Mutable"
        }
```

**Comment by PatMyron** [comment]
wrote a quick [JSON schema](https://json-schema.org/understanding-json-schema/reference/) for the structure of [CloudFormation Resource Specifications](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html) and caught all the `PropertyTypes` missing either `Properties` or `Documentation`:
<details>
  <summary>(click to expand)</summary>

```yaml
    "AWS::SSM::PatchBaseline.PatchStringDate": {
      "PrimitiveType": "String"
    },
    "AWS::Transfer::User.SshPublicKey": {
      "PrimitiveType": "String"
    },
    "AWS::Transfer::Server.Protocol": {
      "PrimitiveType": "String"
    },
    
    "AWS::AppSync::GraphQLApi.AdditionalAuthenticationProviders": {
      "Type": "List",
      "Required": false,
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appsync-graphqlapi-additionalauthenticationproviders.html",
      "ItemType": "AdditionalAuthenticationProvider",
      "UpdateType": "Mutable"
    },
    "AWS::AppSync::GraphQLApi.Tags": {
      "Type": "List",
      "Required": false,
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appsync-graphqlapi-tags.html",
      "ItemType": "Tag",
      "UpdateType": "Mutable"
    },
    "AWS::CodeBuild::Project.FilterGroup": {
      "Type": "List",
      "Required": false,
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codebuild-project-filtergroup.html",
      "ItemType": "WebhookFilter",
      "UpdateType": "Mutable"
    },
    "AWS::CloudWatch::InsightRule.Tags": {
      "Type": "List",
      "Required": false,
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudwatch-insightrule-tags.html",
      "ItemType": "Tag",
      "UpdateType": "Mutable"
    },
    "AWS::Glue::SecurityConfiguration.S3Encryptions": {
      "Type": "List",
      "Required": false,
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-glue-securityconfiguration-s3encryptions.html",
      "ItemType": "S3Encryption",
      "UpdateType": "Mutable"
    },
    "AWS::LakeFormation::DataLakeSettings.Admins": {
      "Type": "List",
      "Required": false,
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lakeformation-datalakesettings-admins.html",
      "ItemType": "DataLakePrincipal",
      "UpdateType": "Mutable"
    },
    
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateActionItemParameterMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateactionitemparametermap.html"
    },
    "AWS::FIS::ExperimentTemplate.ExperimentTemplateActionItemTargetMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-experimenttemplateactionitemtargetmap.html"
    },
    "AWS::FIS::ExperimentTemplate.TagMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fis-experimenttemplate-tagmap.html"
    },
    "AWS::Macie::FindingsFilter.Criterion": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-macie-findingsfilter-criterion.html"
    },
    "AWS::MWAA::Environment.TagMap": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-mwaa-environment-tagmap.html"
    },
    "AWS::SageMaker::DataQualityJobDefinition.Environment": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sagemaker-dataqualityjobdefinition-environment.html"
    },
    "AWS::SageMaker::ModelBiasJobDefinition.Environment": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sagemaker-modelbiasjobdefinition-environment.html"
    },
    "AWS::SageMaker::ModelExplainabilityJobDefinition.Environment": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sagemaker-modelexplainabilityjobdefinition-environment.html"
    },
    "AWS::SageMaker::ModelQualityJobDefinition.Environment": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sagemaker-modelqualityjobdefinition-environment.html"
    },
    "AWS::SageMaker::MonitoringSchedule.Environment": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sagemaker-monitoringschedule-environment.html"
    },
    "AWS::StepFunctions::StateMachine.Definition": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-stepfunctions-statemachine-definition.html"
    },
    "AWS::S3::StorageLens.Encryption": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-storagelens-encryption.html"
    },
```
</details>

**Comment by mrmilosz** [comment]
In case anyone else is struggling with this as I am: this can be monkeypatched with the escape hatch `addPropertyOverride`. For example, if you have your `fis.CfnExperimentTemplate` in a variable called `experimentTemplate` and the action name in a variable called `actionName`, you can get at the parameters of your action with:

```ts
const actionParameters = (experimentTemplate.actions as {
  [key: string]: fis.CfnExperimentTemplate.ExperimentTemplateActionProperty;
})?.[actionName]?.parameters;
```

Then, you can do

```ts
if (actionParameters !== undefined) {
  experimentTemplate.addPropertyOverride(
    `Actions.${actionName}.Parameters`,
    actionParameters
  );
  const duration = (actionParameters as Record<string, any>)["duration"];
  if (duration !== undefined) {
    experimentTemplate.addPropertyOverride(
      `Actions.${actionName}.Parameters.duration`,
      (duration as cdk.Duration).toIsoString()
    );
  }
}
```

Note that because I have a duration in there, I have to convert it to an ISO string (per https://docs.aws.amazon.com/fis/latest/userguide/experiment-template-example.html#cpu-fault-injection) and give it its own override, or else property validation fails with something like `.../duration: expected type: String, found: JSONObject`.

You can do equivalent overrides for the action targets and target resource tags, just without the whole hassle around durations.

**Comment by jwoehrle** [comment]
this seems to be fixed with CDK 1.104.0 🎉 

Thank you !
